ledger/complete/wal/encoding_test.go

```diff
@@ -44,7 +43,7 @@ func TestUpdate(t *testing.T) {
 		operation, stateCommitment, up, err := realWAL.Decode(data)
 		require.NoError(t, err)
 		assert.Equal(t, realWAL.WALUpdate, operation)
-		assert.Equal(t, stateCommitment, ledger.RootHash(hash.DummyHash))
+		assert.Equal(t, stateCommitment, ledger.RootHash(rootHash))
```

Comment:
Hey @fxamacker , could you review my change here? 

I'm not sure why the test case original expect the decoded root hash to be zero. Would like to double check in case I missed anything.

Reply:
@zhangchiqing I think currently `wal.Decode()` returns un-initialized `rootHash` (empty `rootHash`) if operation is `WALUpdate`.  

`rootHash` is only populated in `wal.Decode()` for `WALDelete` operation.

The existing test expecting empty rootHash fails because you changed `wal.Decode()` in this PR.  This can be a breaking change in other use cases as well.

Given this, instead of modifying `wal.Decode()` function, you can use returned `update.RootHash` from `wal.Decode()` for your use case in this PR.

Reply:
OK, I see why it was decided this way. 

There are two types of operation: WALUpdate and WALDelete. 

WALDelete only has the root hash. WALUpdate can be decoded into update, which contains the root hash.

So the operation is WALDelete, than the root hash is returned, and the update is nil; if the operation is WALUpdate, then the decoded update will be returned. So the root hash is only for the case that the operation is WALDelete. 

But I still think in the case the operation is WALUpdate, it should return the root hash. There isn't any code reading the root hash when the operation is WALUpdate. So fixing it would not cause any problem.

This fix would also make my logic simplier, because I don't need to check the returned `update` object, which might be `nil` if the operation is WALDelete. I could just check the root hash and the operation.

Reply:
I took your suggestion and reverted my changes, and also added some commends on the `Decode` method to explain the behavior.